sandbox/cgroup/monitor.go

```diff
@@ -33,122 +33,190 @@ import (
-			folderList = append(folderList, fullPath)
-		} else {
+
+		if !osutil.FileExists(fullPath) {
+			// the path is gone by now
+			delete(newWatch.folders, fullPath)
 			iw.removePath(fullPath)
 		}
 	}
-	if len(folderList) == 0 {
+
+	if len(newWatch.folders) == 0 {
 		newWatch.sendClosedNotification()
 	} else {
-		newWatch.folders = folderList
 		iw.groupList = append(iw.groupList, newWatch)
 	}
+
+	iw.notifyObserver(newWatch)
+
+	logger.Debugf("watches after add: %v", iw.pathList)
 }
 
 func (iw *inotifyWatcher) removePath(fullPath string) {
-	iw.pathList[fullPath]--
-	if iw.pathList[fullPath] == 0 {
-		iw.wd.RemoveWatch(fullPath)
-		delete(iw.pathList, fullPath)
+	cnt, ok := iw.pathList[fullPath]
+
+	if !ok {
+		// path we are not watching
+		return
+	}
+
+	parent := filepath.Dir(fullPath)
+
+	cnt--
+	// we are also keeping references to the parent, see
+	// addWatch about the details
+	iw.pathList[parent]--
+
+	if cnt > 0 {
+		// still references to this path
+		iw.pathList[fullPath] = cnt
+		return
+	}
+
+	logger.Debugf("removing watch for %s", fullPath)
+
+	delete(iw.pathList, fullPath)
+
+	// deal with parent now
+	if iw.pathList[parent] == 0 {
+		if err := iw.wd.RemoveWatch(parent); err != nil {
+			logger.Noticef("cannot remove watch: %v", err)
+		}
+
+		delete(iw.pathList, parent)
 	}
+
+	logger.Debugf("watches after remove: %v", iw.pathList)
 }
 
 // processDeletedPath checks if the received path corresponds to the passed
 // CGroup, removing it from the list of folders being watched in that CGroup if
 // needed. It returns true if there remain folders to be monitored in that CGroup,
 // or false if all the folders of that CGroup have been deleted.
-func (iw *inotifyWatcher) processDeletedPath(watch *groupToWatch, deletedPath string) bool {
-	for i, fullPath := range watch.folders {
-		if fullPath == deletedPath {
-			// if the folder name is in the list of folders to monitor, decrement the
-			// parent's usage counter, and remove it from the list of folders to watch
-			// in this CGroup (by not adding it to tmpFolders)
-			iw.removePath(fullPath)
-			watch.folders = append(watch.folders[:i], watch.folders[i+1:]...)
-			break
-		}
+func (iw *inotifyWatcher) processDeletedPath(watch *groupToWatch, deletedPath string) (keepWatching bool) {
+	if _, ok := watch.folders[deletedPath]; ok {
+		// if the folder name is in the list of folders to monitor, decrement the
+		// parent's usage counter, and remove it from the list of folders to watch
+		// in this CGroup (by not adding it to tmpFolders)
+		iw.removePath(deletedPath)
+		delete(watch.folders, deletedPath)
+
+		iw.notifyObserver(watch)
 	}
 	if len(watch.folders) == 0 {
 		// if all the files/folders of this CGroup have been deleted, notify the
 		// client that it is done.
 		watch.sendClosedNotification()
+
+		iw.notifyObserver(watch)
 		return false
 	}
+
 	return true
 }
 
+func (iw *inotifyWatcher) notifyObserver(w *groupToWatch) {
+	if iw.observeMonitorCb != nil {
```

Comment:
It's unclear how this is synchronized since it seems the notification callback can be called outside of the event loop goroutine. Is this reporting anything on `-race`?

Reply:
the point of adding an observer callback was specifically to have it called within the event loop goroutine